File: November/25/gameoflife.py
class Solution:
    def gameOfLife(self, board):
        m, n = len(board), len(board[0])
        dirs = [(0,1),(0,-1),(1,0),(-1,0),(1,1),(1,-1),(-1,1),(-1,-1)]
        
        for i in range(m):
            for j in range(n):
                live_neighbors = 0
                for dx, dy in dirs:
                    x, y = i + dx, j + dy
                    if 0 <= x < m and 0 <= y < n and board[x][y] in [1,2]:
                        live_neighbors += 1
                # Rule 1 or 3: live -> dead
                if board[i][j] == 1 and (live_neighbors < 2 or live_neighbors > 3):
                    board[i][j] = 2
                # Rule 4: dead -> live
                if board[i][j] == 0 and live_neighbors == 3:
                    board[i][j] = 3
        
        # Finalize board
        for i in range(m):
            for j in range(n):
                if board[i][j] == 2: board[i][j] = 0
                if board[i][j] == 3: board[i][j] = 1
        return board

File: 2025/November/25/gameoflife.cpp
// LeetCode 289: Game of Life
// Explanation:
// 1. In-place state encoding to avoid extra memory.
// 2. 2 represents live->dead, 3 represents dead->live.
// 3. Time: O(m*n), Space: O(1)

#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> gameOfLife(vector<vector<int>>& board) {
        int m = board.size(), n = board[0].size();
        vector<pair<int,int>> dirs{{0,1},{0,-1},{1,0},{-1,0},{1,1},{1,-1},{-1,1},{-1,-1}};
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                int live=0;
                for(auto d: dirs){
                    int x=i+d.first, y=j+d.second;
                    if(x>=0 && x<m && y>=0 && y<n && (board[x][y]==1 || board[x][y]==2)) live++;
                }
                if(board[i][j]==1 && (live<2 || live>3)) board[i][j]=2;
                if(board[i][j]==0 && live==3) board[i][j]=3;
            }
        }
        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++)
                if(board[i][j]==2) board[i][j]=0;
                else if(board[i][j]==3) board[i][j]=1;
        return board;
    }
};

#include <iostream>
int main() {
    Solution s;
    vector<vector<int>> board1{{0,1,0},{0,0,1},{1,1,1},{0,0,0}};
    auto res1 = s.gameOfLife(board1);
    for(auto row: res1){ for(int x: row) cout<<x<<" "; cout<<endl; }

    vector<vector<int>> board2{{1,1},{1,0}};
    auto res2 = s.gameOfLife(board2);
    for(auto row: res2){ for(int x: row) cout<<x<<" "; cout<<endl; }
}
